File: apps/scala/src/main/scala/fizzbuzz/domain/Type.scala
package fizzbuzz.domain

sealed trait FizzBuzzType:
  def generate(number: Int): String

object FizzBuzzType:
  def create(code: Int): FizzBuzzType =
    code match
      case 1 => Type01
      case 2 => Type02
      case 3 => Type03
      case _ => throw new IllegalArgumentException(s"未定義のタイプです: $code")

  case object Type01 extends FizzBuzzType:
    def generate(number: Int): String =
      number match
        case n if n % 15 == 0 => "FizzBuzz"
        case n if n % 3 == 0  => "Fizz"
        case n if n % 5 == 0  => "Buzz"
        case n                => n.toString

  case object Type02 extends FizzBuzzType:
    def generate(number: Int): String = number.toString

  case object Type03 extends FizzBuzzType:
    def generate(number: Int): String =
      number match
        case n if n % 15 == 0 => "FizzBuzz"
        case n                => n.toString

File: apps/clojure/src/fizzbuzz/domain/type.clj
(ns fizzbuzz.domain.type
  (:require [fizzbuzz.domain.model :as model]))

;; ── プロトコル: FizzBuzzType ──────────────────────────────

(defprotocol FizzBuzzType
  "FizzBuzz タイプごとの文字列生成プロトコル"
  (generate-string [this value]
    "FizzBuzzValue から文字列を生成する"))

;; ── タイプ 1: 通常の FizzBuzz ─────────────────────────────

(defrecord FizzBuzzType01 []
  FizzBuzzType
  (generate-string [_this value]
    (let [n (:number value)]
      (cond
        (model/fizz-buzz? n) "FizzBuzz"
        (model/fizz? n) "Fizz"
        (model/buzz? n) "Buzz"
        :else (str n)))))

;; ── タイプ 2: 数字のみ ───────────────────────────────────

(defrecord FizzBuzzType02 []
  FizzBuzzType
  (generate-string [_this value]
    (str (:number value))))

;; ── タイプ 3: FizzBuzz の場合のみ ────────────────────────

(defrecord FizzBuzzType03 []
  FizzBuzzType
  (generate-string [_this value]
    (let [n (:number value)]
      (cond
        (model/fizz-buzz? n) "FizzBuzz"
        (model/fizz? n) "Fizz"
        (model/buzz? n) "Buzz"
        :else ""))))

;; ── タイプ未定義 ─────────────────────────────────────────

(defrecord FizzBuzzTypeNotDefined []
  FizzBuzzType
  (generate-string [_this _value]
    (throw (ex-info "FizzBuzz タイプが未定義です" {:type :not-defined}))))

;; ── ファクトリ関数 ───────────────────────────────────────

(defn create-type
  "タイプ番号から FizzBuzzType インスタンスを生成する"
  [type-num]
  (case type-num
    1 (->FizzBuzzType01)
    2 (->FizzBuzzType02)
    3 (->FizzBuzzType03)
    (->FizzBuzzTypeNotDefined)))
